programs/vote_api/src/vote_transaction.rs

```diff
@@ -30,27 +29,48 @@ impl VoteTransaction {
     /// Fund or create the staking account with tokens
     pub fn new_account(
         from_keypair: &Keypair,
-        vote_account_id: Pubkey,
+        voter_id: Pubkey,
         recent_blockhash: Hash,
         num_tokens: u64,
         fee: u64,
     ) -> Transaction {
-        let create_tx = SystemInstruction::CreateAccount {
-            tokens: num_tokens,
-            space: VoteState::max_size() as u64,
-            program_id: id(),
-        };
-        Transaction::new_with_instructions(
-            &[from_keypair],
-            &[vote_account_id],
-            recent_blockhash,
-            fee,
-            vec![system_program::id(), id()],
-            vec![
-                Instruction::new(0, &create_tx, vec![0, 1]),
-                Instruction::new(1, &VoteInstruction::InitializeAccount, vec![1]),
-            ],
-        )
+        let from_id = from_keypair.pubkey();
+        let space = VoteState::max_size() as u64;
+        TransactionBuilder::new(fee)
+            .push(SystemInstruction::new_program_account(
+                from_id,
+                voter_id,
+                num_tokens,
+                space,
+                id(),
+            ))
+            .push(VoteInstruction::new_initialize_account(voter_id))
+            .sign(&[from_keypair], recent_blockhash)
+    }
+
+    /// Fund or create the staking account with tokens
+    pub fn new_account_with_delegate(
+        from_keypair: &Keypair,
+        voter_keypair: &Keypair,
+        delegate_id: Pubkey,
+        recent_blockhash: Hash,
+        num_tokens: u64,
+        fee: u64,
+    ) -> Transaction {
+        let from_id = from_keypair.pubkey();
+        let voter_id = voter_keypair.pubkey();
+        let space = VoteState::max_size() as u64;
+        TransactionBuilder::new(fee)
+            .push(SystemInstruction::new_program_account(
+                from_id,
+                voter_id,
+                num_tokens,
+                space,
+                id(),
+            ))
+            .push(VoteInstruction::new_initialize_account(voter_id))
+            .push(VoteInstruction::new_delegate_stake(voter_id, delegate_id))
```

Comment:
Question: How does this get past the AccountLoadedTwice error in runtime? 
```
if has_duplicates(&tx.account_keys) {
                error_counters.account_loaded_twice += 1;
                return Err(BankError::AccountLoadedTwice);
            }
```
Is the transactionbuilder doing some magic here?

Reply:
Only a single instruction can't reference the same account twice. It's no problem across multiple instructions.

Reply:
Of maybe you're talking about this magic. :) https://github.com/solana-labs/solana/blob/master/sdk/src/transaction_builder.rs#L15

Reply:
When the same account is used by multiple instructions, it is only added to tx.account_keys once, and identified by the instruction indexes.
The AccountLoadedTwice error is only triggered is the same account is actually added to tx.account_keys twice.

Reply:
(Oops, sorry for the overkill. Should have refreshed!)

Reply:
Yup that's how I expected it to work, makes sense now. Thanks!